README.md
# renren-chrome
Chrome extension for RenRen web service.

You can download it and see how it looks like from https://chrome.google.com/webstore/detail/%E4%BA%BA%E4%BA%BA-for-chrome/nkkfldahifjgamnkimpbabokkgiichja

changelog.txt
更新历史:
#2012-4-17
  >添加了分享新鲜事的评论功能

#2011-12-23
  >修正了由于RenrenAPI参数变化引起得状态重复出现的问题
#2011-10-7
  >修正了由于RenrenAPI参数变化引起得状态重复出现的问题
  >添加了从网页选择一段文本后右键直接当作状态发布的功能
  >添加了状态的转发功能，暂不能添加评论转发
#2011-8-30 v1.04
  >通过了人人审核，用户不再有数量限制，之前困扰多时的问题应该得以修复了
#2011-7-11 v1.0.4
  >修正了有时分享中缩略图过大的问题
  >修正了每次启动Chrome必须点击2次图标才能启动插件的问题
  >修正了状态中地理位置显示的错误（目前仅在个人状态中有效，因为新鲜事中尚未返回地理位置信息）
#2011-6-30 v1.0.3
  >修正了分享网页和发状态偶尔出现的无响应现象
#2011-6-24 v1.0.2 beta
  >修正了获取oAuth认证时的字符串错位
#2011-6-21 v1.0.1 beta
  >添加对留言的回复功能
#2011-6-19 v1.0 beta
  >主要功能完成
  >第一版本发布

js/pages.js
/*
 * 加载各个页面时调用的函数
 * 作者: Zhang Fan
 * 写于: 2011-6-12
 */

//======================================通知页面（暂不支持）start================================
function showNotificationsPage(ctrl){
	localStorage.lastTab = ctrl.id;
	// 切换当前页面icon的式样
	Func.selectCurrentPage(ctrl);
	// 隐藏其他页面
	Func.hidePages();
	document.getElementById("notifications").style.display = "block";
	// 显示新通知
	// sorry, Renren暂没有获取通知的api，WTF
}
//======================================通知页面（暂不支持）end================================

//======================================新鲜事页面start================================
function showNewsfeedPage(ctrl) {
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	// 显示状态输入框
	document.getElementById("newpost").style.display = "block";
	// 显示页面
	document.getElementById("newsfeed").style.display = "block";
	// 加载旧页面内容
	try{
		if(localStorage.newsfeedCache){
			var cache = JSON.parse(localStorage.newsfeedCache);
			// 输出缓存内容
			document.getElementById("newsfeed").innerHTML = writeNewsfeed(cache);
		}
	}catch(e){}
	// 获取新内容
	RR.api({
		method: "feed.get",
		  type: "10,11,20,21,22,23,30,31,32,33,34,35,36,50,51",
		 count: "50"
	},
	function(res){
		if(!res.error_code){
			// 正确获取内容，写入页面
			if(res.length != 0){
				document.getElementById("newsfeed").innerHTML = writeNewsfeed(res);
				localStorage.newsfeedCache = JSON.stringify(res);
			}else{
				console.log("There is nothing published.");
			}
		}else{
			console.log(res.error_code + " : " + res.error_msg);
		}
	});
}
//======================================新鲜事页面end================================

//======================================状态页面start================================
function showStatusPage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("newpost").style.display = "block";
	
	document.getElementById("status").style.display = "block";
	if(localStorage.statusCache){
		var cache = JSON.parse(localStorage.statusCache);
		document.getElementById("status").innerHTML = writeNewsfeed(cache);
	}
	RR.api({
		method: "feed.get",
		type: "10"
	},function(res){
		if(!res.error_code){
			if(res.length != 0){
				document.getElementById("status").innerHTML = writeNewsfeed(res);
				localStorage.statusCache = JSON.stringify(res);
			}else{
				console.log("There is nothing.");
			}
		}else{
			console.log(res.error_code + " : " + res.error_msg);
		}
	});
}
//======================================状态页面end================================

//======================================相册图片页面start================================
function showAlbumPage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("newpost").style.display = "block";
	
	document.getElementById("album").style.display = "block";
	if(localStorage.albumCache){
		var cache = JSON.parse(localStorage.albumCache);
		document.getElementById("album").innerHTML = writeNewsfeed(cache);
	}
	RR.api({
		method: "feed.get",
		type: "30,34"
	},function(res){
		if(!res.error_code){
			if(res.length != 0){
				document.getElementById("album").innerHTML = writeNewsfeed(res);
				localStorage.albumCache = JSON.stringify(res);
			}else{
				console.log("There is nothing.");
			}
		}else{
			console.log(res.error_code + " : " + res.error_msg);
		}
	});
}
//======================================相册图片页面end================================

//======================================日志页面start================================
function showBlogPage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("newpost").style.display = "block";
	document.getElementById("blog").style.display = "block";
	if(localStorage.blogCache){
		var cache = JSON.parse(localStorage.blogCache);
		document.getElementById("blog").innerHTML = writeNewsfeed(cache);
	}
	RR.api({
		method: "feed.get",
		type: "20"
	},function(res){
		if(!res.error_code){
			if(res.length != 0){
				document.getElementById("blog").innerHTML = writeNewsfeed(res);
				localStorage.blogCache = JSON.stringify(res);
			}else{
				console.log("There is nothing.");
			}
		}else{
			console.log(res.error_code + " : " + res.error_msg);
		}
	});
}
//======================================日志页面end================================

//======================================分享页面start================================
function showSharePage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("newpost").style.display = "block";
	
	document.getElementById("share").style.display = "block";
	if(localStorage.shareCache){
		var cache = JSON.parse(localStorage.shareCache);
		document.getElementById("share").innerHTML = writeNewsfeed(cache);
	}
	RR.api({
		method: "feed.get",
		type: "21,32,33,50,51"
	},function(res){
		if(!res.error_code){
			if(res.length != 0){
				document.getElementById("share").innerHTML = writeNewsfeed(res);
				localStorage.shareCache = JSON.stringify(res);
			}else{
				console.log("There is nothing.");
			}
		}else{
			console.log(res.error_code + " : " + res.error_msg);
		}
	});
}
//======================================分享页面end================================

//======================================公共主页页面start================================
function showPublicPage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("newpost").style.display = "block";
	
	document.getElementById("public").style.display = "block";
	if(localStorage.publicCache){
		var cache = JSON.parse(localStorage.publicCache);
		document.getElementById("public").innerHTML = writeNewsfeed(cache);
	}
	RR.api({
		method: "feed.get",
		type: "11,22,23,31,35,36,53,54"
	},function(res){
		if(!res.error_code){
			if(res.length != 0){
				document.getElementById("public").innerHTML = writeNewsfeed(res);
				localStorage.publicCache = JSON.stringify(res);
			}else{
				console.log("There is nothing.");
			}
		}else{
			console.log(res.error_code + " : " + res.error_msg);
		}
	});
}
//======================================公共主页页面end================================

//======================================自己的状态页面start================================
function shwoMyfeedPage(ctrl) {
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("newpost").style.display = "block";
	document.getElementById("myfeed").style.display = "block";
	// 加载旧页面内容
	try{
		if(localStorage.myfeedCache){
			var cache = JSON.parse(localStorage.myfeedCache);
			// 套入模板
			document.getElementById("myfeed").innerHTML = writeMyfeed(cache);
		}
	}catch(e){}
	// 获取新内容
	RR.api({
		method: "status.gets",
		count: "15"
		},function(res){
			if(!res.error_code){
				// 正确获取内容，写入页面并缓存
				document.getElementById("myfeed").innerHTML = writeMyfeed(res);
				localStorage.myfeedCache = JSON.stringify(res);
			}else{
				console.log(res.error_code + " : " + res.error_msg);
			}
	});
}
//======================================自己的状态页面end================================

/*//======================================留言板页面start（暂不支持）================================
function showMycommentsPage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("mycomments").style.display = "block";
	if(localStorage.mycommentsCache){
		var cache = JSON.parse(localStorage.mycommentsCache);
	}
	RR.api({
		method: ""
	},function(res){
		
	});
}
//======================================状态页面end================================*/

/*//====================================== 站内信页面start（暂不支持）================================
function showInboxPage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	document.getElementById("inbox").style.display = "block";
}*/
//=========================================站内心页面end================================

//======================================好友页面end================================
/* 每次加载页面第一次要重新读取好友列表（可能期间添加了新好友） */
var loadedFriends = false;
function showFriendsPage(ctrl){
	localStorage.lastTab = ctrl.id;
	Func.selectCurrentPage(ctrl);
	Func.hidePages();
	// hide发布类型列表
	document.getElementById("type_dropdown").style.display = "none";
	document.getElementById("friends").style.display = "block";
	document.getElementById("friendsort").style.display = "block";
	if(loadedFriends){
		if(localStorage.friendsCache){
			if(document.getElementById("friendsort_text").value != ""){
				document.getElementById("friendsort_text").onkeyup();
			}else{
				var cache = JSON.parse(localStorage.friendsCache);
				document.getElementById("friends").innerHTML = writeFriends(cache);
			}
		}
	}else{
		RR.api({method: "friends.getFriends"}, function(res){
			if(!res.error_code){
				// 正确获取内容
				loadedFriends = true;
				// 按名字排序
				res.sort(function(a,b){
					var x = a.name;
					var y = b.name;
					return x.localeCompare(y);
				});
				/* 根据好友搜索框显示内容
				if(document.getElementById("friendsort_text").value != ""){
					document.getElementById("friendsort_text").onkeyup();
				}else{
					document.getElementById("friendsort_text").innerHTML = "";
					alert(JSON.stringify(res));
				}*/
				document.getElementById("friends").innerHTML = writeFriends(res);
				localStorage.friendsCache = JSON.stringify(res);
			}else{
				console.log(res.error_code + " : " + res.error_msg);
			}
		});
	}
}
/* 查找好友输入框，随文字变化改变页面内容 */
function searchFriends(ctrl){
	// 从加载过的数据中按输入字符查找
	var friends = JSON.parse(localStorage.friendsCache),
		matches = [],
		text = ctrl.value,
		regx = new RegExp("\\b" + text.replace(/\W/g, '').split('').join('.{0,3}'), "i");
	for(var friend in friends){
		try{
			if(friends[friend].name.match(regx)){
				matches.push(friends[friend]);
			}
		}catch(e){}
	}
	matches.sort(function(a,b){
		var x = a.name;
		var y = b.name;
		return x.localeCompare(y);
	});
	// 刷新页面
	document.getElementById("friends").innerHTML = writeFriends(matches);	
}
//======================================好友页面end================================

//======================================常用自定义函数==============================
var Func = {
	// 只高亮当前页面icon
	selectCurrentPage : function(ctrl){
		var siblings = ctrl.parentNode.childNodes;
		for(var i = 0; i < siblings.length; i++){
			if(typeof siblings[i] != "string"){
				siblings[i].className = "";
			}
		}
		ctrl.className = "selected";
	},
	// 隐藏其他页面
	hidePages : function(){
		var pages = document.getElementsByClassName("page");
		for(var i = 0; i < pages.length; i++){
			pages[i].style.display = "none";
		}
	}
};

js/renren.js
/*
 * 人人API调用的封装
 * 作者: Zhang Fan
 * 作于: 2011-6-12
 */

var RR = {
	// Renren API的URL和参数
	api_id : "140612",
	api_key : "e9a919ae27874f7a8a10b8186b6a4812",
	URL : {
		api: "http://api.renren.com/restserver.do",
		oauth: "https://graph.renren.com/oauth/authorize",
		redirectUrl: "http://graph.renren.com/oauth/login_success.html"
	},
	
  	// 检查localStorage中是否存有access_token以及是否过期,如果正常返回access_token
	token : function() {
		if(localStorage.rr_access_token && localStorage.rr_access_token != "false" && localStorage.rr_access_token != "undefinied"
			&& localStorage.rr_token_expire && localStorage.rr_token_expire != 0 && localStorage.rr_token_expire > Date.now()){
			return localStorage.rr_access_token;
		}else{
			return false;
		}
	},

	// 发送oAuth2.0请求
	login : function() {
		//请求的权限
		var scopes = ["read_user_status","read_user_blog","read_user_feed","read_user_share",
		             "read_user_photo","read_user_album","read_user_checkin","read_user_comment",
		             "publish_feed","publish_share","publish_checkin","publish_comment",
		             "status_update","photo_upload","operate_like"];
		var scope = "";
		for(var i = 0; i < scopes.length; i++){
			if(i == 0){
				scope += scopes[i];
			}
			scope += " " + scopes[i];
		}
		//生成请求URL
		var url = this.URL.oauth + "?client_id=" + this.api_key + "&response_type=token"
			+ "&redirect_uri=" + this.URL.redirectUrl
			+ "&scope=" + scope;
		//弹出屏幕，居中
		var w = 580;
		var h = 360;
		var l = (screen.width - w) / 2;
        var t = (screen.height - h) / 2;
        var p = 'width=' + w + ', height=' + h + ', top=' + t + ', left=' + l;
        p += ', toolbar=no, scrollbars=no, menubar=no, location=no, resizable=no';
		window.open(url,"Renren connection",p);
	},
	
	// 去除对此应用的oAuth认证
	logout : function(callback){
		// TODO 移除Renren服务器上的认证信息（Renren未提供）
		callback(true);
	},

	// 调用Renren rest API的入口函数
  	api : function(options, callback) {
  		
		// 预设几个API必须的参数
		options["access_token"] = this.token();
		options["format"] = "JSON";
		options["v"] = "1.0";
		options["call_id"] = Date.now();
		
		// 按字典顺序排序params
		var order = [];
		for(var key in options){
			order.push(key);
		}
		order.sort();
		
  		// 用md5计算要求的签名sig
		md5str = "";
		for(var i = 0; i < order.length; i++){
			if(order[i] != "upload"){
				md5str += order[i] + "=" + options[order[i]];
			}
		}
		var sig = MD5(md5str + "349181b8d9cb41ee82d00f3b0d3dc27d");

		// 组合请求url
		var url = this.URL.api + "?" + order[0] + "=" + options[order[0]];
		for(i = 1; i < order.length; i++){
			url += "&" + order[i] + "=" + encodeURIComponent(options[order[i]]);
		}
		url += "&sig=" + sig;
		
		// 显示 ajax loading 图标
		// Fixed by Fortable1999
		if(document.getElementById("loader")){
			document.getElementById("loader").style.display = "block";
		}
		else{
			console.log("Error in getElementByID: 'loader'");
		}
  		// XMLHttp请求
  		var xhr = new XMLHttpRequest();
  		// 一般情况下Renren要求用POST请求
		xhr.open("POST", url, true);
		xhr.onreadystatechange = function() {
			if (xhr.readyState == 4) {
				if (xhr.status == 200){
					// 返回response
					// alert(xhr.responseText);
					callback(JSON.parse(xhr.responseText));
					// 隐藏 AJAX loading 图标
					if(document.getElementById("loader")){
						document.getElementById("loader").style.display = "none";
					}
				}else{
					console.log("XMLHttpRequest error.");
					// 隐藏 AJAX loading 图标
					if(document.getElementById("loader")){
						document.getElementById("loader").style.display = "none";
					}
				}
			}
		};
		xhr.send();
  	}
};

js/template.js
/*
 * 关于各个页面的模板，即初始页面加载使用的模板
 * 作者: Zhang Fan
 * 写于: 2011-6-15
 * 最新修改于： 2012-4-17
 */
//====================================== 自己的状态start ======================================
function writeMyfeed(total){
	var result = "";
	for(var one in total){
		result += writeOneMyfeed(JSON.parse(localStorage.userInfo),total[one]);
	}
	return result;
}
function writeOneMyfeed(user, one){
	var params = [];
	// 如果含有地址信息，显示地理位置
	if(one.place){
		params.push(' &middot; 在 <a href="http://places.renren.com/web/PoiInfo?id=' + one.place.lbs_id + '&type=2&statID=' + one.status_id + '&level=" target="_blank">' + one.place.name + '</a>');
	}else{
		params.push(' ');
	}
	return '<li id="' + one.status_id + '">'
			+ '<img class="post_actorpic" src="' + user.tinyurl + '" style="width:50px;height:50px"/>'
			+ '<div class="post_content">'
				+ '<div class="post_actors">'
					+ '<a href="http://www.renren.com/profile.do?id=' + user.uid + '" target="_blank">' + user.name + '</a>'
				+ '</div>'
				+ '<div class="message">' + formatMessage(one.message)
					+ '<div style="overflow:hidden"></div>'
				+ '</div>'
				+ '<div class="options">'
					+ '<span title="' + one.time + '" class="timestring">' + prettyDate(one.time) + '</span>' + params[0]
				+ '</div>'
				+ '<div class="post_comments">'
					+ '<div>'
						+ '<div class="post_comment"' + (one.comment_count == 0?' style="display:none"':'') + '>'
							+ '<a href="#' + one.status_id + '" onclick="getComments(this,' + one.uid + ',' + one.status_id + ',10)" class="view_all_comments">查看全部' + (one.comment_count > 50?(one.comment_count+'条回复中最新的50条'):(one.comment_count+'条回复')) + '</a>'
						+ '</div>'
					+ '</div>'
					+ '<div class="post_comment">'
						+ '<div class="comment_form" title="' + one.status_id + '">'
							+ '<input type="text" class="comment_input" placeholder="点击添加回复" onfocus="showCommentBox(this)" onblur="hideCommentBox(this)" />'
							+ '<input type="button" class="button smallbutton" value="回复" style="display:none" onclick="addComment(this,' + one.uid + ',' + one.status_id + ',10)" />'
						+ '</div>'
					+ '</div>'
				+ '</div>'
			+ '</div>'
		+ '</li>';
}
//========================================自己的状态end==================================

//=========================================新鲜事start==================================
/* 新鲜事模板 */
function writeNewsfeed(total){
	var result = "";
	for(var i in total){
		result += writeOneNewsfeed(total[i]);
	}
	return result;
}
function writeOneNewsfeed(one){
	var params = [];
	// 如果不是状态，可能没有message
	if(one.message && one.message != ""){
		params["message"] = formatMessage(one.message);
	}
	// 如果含有地址信息，显示地理位置
	if(one.place){
		params["place"] = ' &middot; 在 <a href="'+ one.place.url +'">' + one.place.name + '</a>';
	}
	// 如果是分享的链接带有图片或是视频信息
	if(one.attachment && one.attachment[0]){
		params["attachment"] = '';
		if(one.attachment[0].href)
			params["attachment"] += '<a href="' + one.attachment[0].href + '" target="_blank">';
			if(one.attachment[0].src){
				params["attachment"] += '<div class="post_pictures"><img src="' + one.attachment[0].src + '" class="post_picture" /></div>';
			}
			params["attachment"] += '</a>';
	}
	// 如果是日志，分享链接，图片或视频，则带有题目，link和概要
	if(one.title){
		if(one.href){
			params["title"] = '<a href="' + one.href + '" target="_blank">' + one.title + '</a>';
		}else{
			params["title"] = one.title;
		}
	}
	// 概要
	if(one.description){
		params["description"] = '<p>' + one.description + '</p>';
	}
	// 支持转发的类型，显示（转发状态，分享其他分享）
	if(one.feed_type == 10){
		params["forward"] = ' &middot; <a href="#null" onclick="forwardStatus(' + one.actor_id + ',' + one.source_id + ')">转发</a>';
	}
	// 提供like,unlike操作此条目
	if(one.likes && one.likes.user_like == 0){
		// 不支持like操作的不显示
		if(parseInt(one.feed_type/10) == 5){
			params["userLike"] = ' &middot; <a href="#">喜欢</a>';			
		}
	}else{
		params["userLike"] = ' &middot; <a href="#">取消喜欢</a>';
	}
	// 有人表示喜欢的话，显示
	if(one.likes && one.likes.total_count > 0){
		params["likes"] = '<div class="post_likes">'
			+ formatLikes(one.likes)
			+ '</div>';
	}
	// 如果存在评论
	if(one.comments && one.comments.count > 0){
		params["comments"] = "";
		if(one.comments.count > 2){
			var cmt = [];
			for(var i in one.comments.comment){
				cmt[i] = showDefNewsfeedComment(one.comments.comment[i]);
			}
			// 多余2条时是两条中间夹着的式样
			params["comments"] = cmt[0]
				+ '<div class="post_comment">'
					+ '<a class="view_all_comments" href="#' + one.post_id +'" onclick="getComments(this,' + one.actor_id + ',' + one.source_id +',' + one.feed_type + (one.attachment && one.attachment[0] && one.attachment[0].media_id?(','+one.attachment[0].media_id):'') + ')">查看全部' + (one.comments.count > 50?(one.comments.count+'条回复中最新的50条'):(one.comments.count+'条回复')) + '</a>'
				+ '</div>'
				+ cmt[1];
		}else{
			var cmt = [];
			for(var i in one.comments.comment){
				cmt[i] = showDefNewsfeedComment(one.comments.comment[i]);
			}
			// 少于两条就都显示
			for(var i in cmt){
				params["comments"] += cmt[i];
			}
		}
	}
	// 整合一个newsfeed
	return '<li id="' + one.post_id + '">'
			+ '<img class="post_actorpic" src="' + one.headurl + '" style="width:50px;height:50px" />'
			+ '<div class="post_content">'
				+ '<div class="post_actors">'
					+ '<a href="http://www.renren.com/profile.do?id=' + one.actor_id + '" target="_blank">' + one.name + '</a>  ' + (one.prefix && one.feed_type != 10?one.prefix:'')
				+ '</div>'
				+ '<div class="message">'
					+ (one.message?params["message"]:'')
					+ '<div style="overflow:hidden">'
						+ (one.attachment && one.attachment[0]?params["attachment"]:'')
						+ (one.title && one.feed_type != 10?params["title"]:'')
						+ (one.description?params["description"]:'')
					+ '</div>'
				+ '</div>'
				+ '<div class="options">'
					+ '<span class="timestring" title="' + one.update_time + '">' + prettyDate(one.update_time) + '</span>'
					+ (one.place?params["place"]:'')
					+ (one.feed_type == 10?params["forward"]:'')
					+ ((false && one.likes.user_like != "undefined" && parseInt(one.feed_type/10) == 5)?params["userLike"]:'')
				+ '</div>'
				+ '<div class="post_comments">'
					+ (one.likes && one.likes.total_count > 0?params["likes"]:'')
					+ '<div>'
						+ (one.comments && one.comments.count > 0?params["comments"]:'')
					+ '</div>'
					/*+ ((parseInt(one.feed_type/10) != 5 && one.feed_type != 21 && one.feed_type != 23 && one.feed_type != 32 && one.feed_type != 33 && one.feed_type != 36)?*/
					+'<div class="post_comment">'
						+ '<div class="comment_form" title="' + one.post_id + '">'
							+ '<input type="text" class="comment_input" placeholder="点击添加回复" onfocus="showCommentBox(this)" onblur="hideCommentBox(this)" />'
							//+ (one.feed_type == 10?'<input type="checkbox" value="' + one.source_id + '">同时转发':'')
							+ '<input type="button" class="button smallbutton" value="回复" style="display:none" onclick="addComment(this,' + one.actor_id + ',' + one.source_id +',' + one.feed_type + (one.attachment && one.attachment[0] && one.attachment[0].media_id?(','+one.attachment[0].media_id):'') + ')" />'
						+ '</div>'
					+ '</div>'/*:'')*/
				+ '</div>'
			+ '</div>'
		+ '</li>';

}
//===========================================新鲜事end=========================================

//===========================================好友start========================================
/* 好友列表页面模板 */
function writeFriends(total){
	var result = "";
	for(var i in total){
		result += writeOneFriend(total[i]);
	}
	return result;
}
function writeOneFriend(one){
	return '<li>'
			+ '<img class="post_actorpic" src="' + one.tinyurl + '" style="width:36px;height:36px" />'
			+ '<div class="post_content">'
				+ '<a href="http://www.renren.com/profile.do?id=' + one.id + '" target="_blank">' + one.name +'</a>'
				+ '<div class="options">'
					+ '<a href="http://www.renren.com/profile.do?id=' + one.id +'" target="_blank">前往个人主页</a> &middot; '
					+ '<a href="http://gift.renren.com/show/gift/home?fid=' + one.id + '" target="_blank">送TA礼物</a>'
				+ '</div>'
			+ '</div>'
		+ '</li>';
}
//=============================================好友end=======================================

//============================================回复模板 start===============================================
/* 
 * 显示新鲜事默认返回的回复（最多2条的） 
 * 不能改了，因为新鲜事返回的默认回复参数
 */
function showDefNewsfeedComment(one){
	return '<div class="post_comment">'
			+ '<img class="post_actorpic" src="' + one.headurl + '"/>'
			+ '<div class="post_comment_content">'
				+ '<a href="http://www.renren.com/profile.do?id="' + one.uid + '" target="_blank">' + one.name + '</a>  '
				+ one.text.substr(1,one.text.length-2).replace(/\\/g,"") // 新鲜事默认值中所有"前都了一个反斜杠
				+ '<div class="options">'
					+ '<span class="timestring" title="' + one.time + '">' + prettyDate(one.time) + '</span>'
					+ (one.uid != localStorage.userId?('  &middot;  <a href="#null" onclick="replyComment(this,' + one.uid + ',' + "'" + one.name + "'" + ')">回复</a>'):'')
				+ '</div>'
			+ '</div>'
		+ '</div>';
}

/* 显示点击后获取的全部回复 
 * total内容按时间由新到旧
 * 逆序输出
 */
function showComments(total, type){
	var com = "";
	for(var i = (total.length - 1); i >= 0 ; i--){
		com += generateOneComment(total[i], type);
	}
	return com;
}

/* 显示点击后获取的全部回复
 * total内容按时间由旧到新
 * 逆序输出（有的api不支持order）
 */
function showCommentsOld(total, type){
	var com = "";
	for(var i in total){
		com += generateOneComment(total[i], type);
	}
	return com;
}

/* 生成一个回复的模板 */
function generateOneComment(one, type){
	//统一参数表示
	var params = new Array(); 
	if(type == 1){
		// status
		params["comment_id"] = one.comment_id;
		params["uid"] = one.uid;
		params["name"] = one.name;
		params["tinyurl"] = one.tinyurl;
		params["time"] = one.time;
		params["content"] = one.text;
	}else if(type == 2){
		// blog
		params["comment_id"] = one.id;
        params["uid"] = one.uid;
        params["name"] = one.name;
        params["tinyurl"] = one.headurl;
        params["time"] = one.time;
        params["content"] = one.content;
	}else if(type == 3){
		// ablum
		params["comment_id"] = one.comment_id;
        params["uid"] = one.uid;
        params["name"] = one.name;
        params["tinyurl"] = one.headurl;
        params["time"] = one.time;
        params["content"] = one.text;
	}else if(type == 5){
		// share
		params["comment_id"] = one.id;
        params["uid"] = one.uid;
        params["name"] = one.name;
        params["tinyurl"] = one.headurl;
        params["time"] = one.time;
        params["content"] = one.content;
	}
	return '<div class="post_comment" id="' + params.comment_id + '" >'
			+ '<img class="post_actorpic" src="' + params.tinyurl + '" />'
			+ '<div class="post_comment_content">'
				+ '<a href="http://www.renren.com/profile.do?id="' + params.uid + '" target="_blank">' + params.name + '</a>  ' + params.content
				+ '<div class="options">'
					+ '<span title="' + params.time + '" class="timestring">' + prettyDate(params.time) + '</span>'
					+ (params.uid != localStorage.userId?('  &middot;  <a href="#null" onclick="replyComment(this,' + params.uid + ',' + "'" + params.name + "'" + ')">回复</a>'):'')
				+ '</div>'
			+ '</div>'
		+ '</div>';
}
//============================================回复模板 end===============================================

//============================================模板样式补充函数=========================================
// 替换文字url为链接
function formatMessage(msg){
	msg = msg.replace(
			/((https?\:\/\/|ftp\:\/\/)|(www\.))(\S+)(\w{2,4})(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/gi,
			function(url){return '<a href="'+ url +'" target="_blank">'+ url +'</a>';}
			);
	return msg.replace(/<3/g,"&hearts;").replace(/[\n\r]/g,"<br />");
}

// 修改喜欢操作为fb的样式
function formatLikes(likes) {
	var out = "";
	if(likes.user_like == 1){
		out = "你";
	}
	if(likes.total_count >= 2) {
		out += "和另外" + (likes.total_count - 1) + "人表示喜欢.";
	} else if(likes.count == 1) {
		out += "表示很喜欢.";
	} else {
		out += "没人表示喜欢.";
	}
	return out;
}

// 显示发布时间到现在的时间差
// JavaScript Pretty Date
// Copyright (c) 2008 John Resig (jquery.com)
// Licensed under the MIT license.
// Takes an ISO time and returns a string representing how
// long ago the date represents.
function prettyDate(time){
	
	var m = time.match(/^\s*(\d{4})-(\d\d)-(\d\d)\s+(\d\d):(\d\d):*(\d\d)*\s*$/);
	// 防止有些传递时间没有秒
	if(typeof m[6]  == "undefined"){
		m[6] = 0;
	}

	var date;
	try {
		date = new Date(m[1],m[2]-1,m[3],m[4],m[5],m[6]);
	}catch(e) {
		date = new Date(time);
	}

	diff = (((new Date()).getTime() - date.getTime()) / 1000),
	day_diff = Math.floor(diff / 86400);

	if ( isNaN(day_diff) || day_diff < 0 || day_diff >= 31 )
		return "至少一个月以前";

	return day_diff == 0 && (
			diff < 60 && "刚刚更新" ||
			diff < 120 && "1分钟前" ||
			diff < 3600 && Math.floor( diff / 60 ) + "分钟前" ||
			diff < 7200 && "1小时前" ||
			diff < 86400 && Math.floor( diff / 3600 ) + "小时前") ||
			day_diff == 1 && "昨天" ||
			day_diff < 7 && day_diff + "天前" ||
			day_diff < 31 && Math.ceil( day_diff / 7 ) + "星期前";
}